### PR TITLE
Support Bed Distance Sensor

### PR DESCRIPTION
**Pull request by markniu**
BDsensor is a high resolution inductive Bed Distance Sensor for auto bed leveling, it can measure the distance from bed to nozzle with distance resolution 0.01mm in real time.

More detail : https://github.com/markniu/Bed_Distance_sensor
Configure: https://github.com/markniu/Bed_Distance_sensor/wiki/Installing-for-Klipper

add the following into the printer.cfg and change the pins: sda and slc to yours
```
[BDsensor] 
sda_pin: PB0
scl_pin: PB1
delay: 800000
z_offset:0
```
and for display the real time distance value from the BDsensor on the LCD12864 screen, just add the following line to the [display] of the printer.cfg

`BD_display:`

you can also get the distance value by send gcode command `M102   S-2 `

this code has been tested on some printers, like Ender3 with stm32f103(creality V4.2.7)board, mks nano v3

Signed-off-by:Mark
<niujl123@sina.com>

### CODE REVIEW

File: src/BD_sensor.h [hunk 1]
```diff
@@ -0,0 +1 @@
+uint16_t Get_Distane_data(void);
```
Comment (thijstriemstra) [inline src/BD_sensor.h:1]: typo (distance)

**Comment by JamesH1978** [comment]
Please refer to https://github.com/Klipper3d/klipper/blob/master/docs/CONTRIBUTING.md and make sure you have included a signed-off-by line.

Thanks
James

File: src/BD_sensor.h (hunk 1 above)
Comment (markniu) [reply src/BD_sensor.h]: thanks, that header file has been deleted for it is unused now.

File: klippy/extras/BDsensor.py [hunk 2]
```diff
@@ -0,0 +1,288 @@
+#import libraries
```
Comment (thijstriemstra) [inline klippy/extras/BDsensor.py:1]: remove unused imports/dead code and add license header

**Comment by thijstriemstra** [comment]
Unfortunately I, or anyone else contributing to this repository, have access to such a device so I can't test it. Not sure what the policy for such new devices is, I assume a reviewer will have to test the actual device before a new sensor can be added to Klipper @JamesH1978?

File: klippy/extras/BDsensor.py [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+    max_cruise_v2 = dist * accel
+    if max_cruise_v2 < speed**2:
+        speed = math.sqrt(max_cruise_v2)
+    accel_t = speed / accel
+    accel_decel_d = accel_t * speed
+    cruise_t = (dist - accel_decel_d) / speed
+    return axis_r, accel_t, cruise_t, speed
+
+
+# I2C BD_SENSOR
```
Comment (thijstriemstra) [inline klippy/extras/BDsensor.py:27]: you should move this info to a `MCU_I2C_BD` class docstring instead.

**Comment by markniu** [comment]
> Unfortunately I, or anyone else contributing to this repository I think, have access to such a device so I can't test it. Not sure what the policy for such new devices is, I assume a reviewer will have to test the actual device before a new sensor can be added to Klipper @JamesH1978?

I'd like to send some free sensors out to someone for test or review, just send shipping address to my email.

**Comment by JamesH1978** [comment]
I'm not a reviewer for this kind of thing, I review printer/board config mainly :)

Generally a reviewer does not need to have the actual equipment, but they will need some real world proof that it works. And that can be just from the person submitting a PR. It is not as if we have every printer and every option sent to us for any kind of verification. That would be untenable, as most of us do not do this full time.

I assume this will be reviewed by Kevin or Arksine, and it just depends when they have time to look at your code and review it.

Thanks
James

**Comment by thijstriemstra** [comment]
> I'd like to send some free sensors out to someone for test or review, just send shipping address to my email.

I wouldn't mind help testing, if you have a spare board, my email address is info at collab.nl

File: klippy/extras/BDsensor.py [hunk 4]
```diff
@@ -0,0 +1,266 @@
+# Bed leveling sensor BDsensor(Bed Distance sensor)
+# https://github.com/markniu/Bed_Distance_sensor
+# Copyright (C) 2023 Mark yue <niujl123@sina.com>
+# This file may be distributed under the terms of the GNU GPLv3 license.
+
+import chelper
+import mcu,math
+from . import probe
+
+# Calculate a move's accel_t, cruise_t, and cruise_v
```
Comment (thijstriemstra) [inline klippy/extras/BDsensor.py:10]: move this into function docstring instead

**Comment by Printzp** [comment]
I'd be willing to contribute to testing for this. I now have the sensor, and am now struggling getting it to work. So, any data I can provide I'm more than willing to do so.

**Comment by markniu** [comment]
> 



> I'd be willing to contribute to testing for this. I now have the sensor, and am now struggling getting it to work. So, any data I can provide I'm more than willing to do so.

Thanks, please feel free to contact me if you have any problem while using it.

**Comment by markniu** [comment]
> collab.nl

seems that I can not open your websit https://[www.collab.nl](https://www.collab.nl/)/    return HTTP ERROR 502

**Comment by thijstriemstra** [comment]
@markniu I'll look into it but this does not affect my email.

**Comment by mcmasterp** [comment]
I have these sensor in hand and can do whatever testing you need. This sensor idea is incredible and the future for hobby level printers at least if not more. It's definitely a great thing for klipper to support asap if possible

**Comment by Zchulski** [comment]
I have one of these sensors as well but without this getting merged into the main repo i'm struggling to get klipper to even work correctly.  I would love to be able to test as well but unless I can get klipper to work correctly on his fork this is a difficult task

**Comment by Printzp** [comment]
> I have one of these sensors as well but without this getting merged into the main repo i'm struggling to get klipper to even work correctly. I would love to be able to test as well but unless I can get klipper to work correctly on his fork this is a difficult task

Yes, sadly I'm in the same boat. I'm excited for it to get added though.

**Comment by Sineos** [comment]
> you can also get the distance value by send gcode command `M102 S-2 `

@KevinOConnor would need to chime in but in general Klipper does not introduce or modify additional Gcodes but establishes new functionality via "human readable" commands.
For an example see the `probe` commands https://www.klipper3d.org/G-Codes.html#probe or other commands in the linked document.

**Comment by Zchulski** [comment]
I finally got the sensor to communicate with klipper but sadly I don't see any useful data coming from klipper.  I cannot get this to calibrate correctly.  there is something going on between the sensor and klipper that is preventing this to spit out useful data.  I have followed marks instructions to the letter a handful of times and even tried to alter the instructions a little bit to try and get different results but to no avail.

**Comment by Zchulski** [comment]
> I'm not a reviewer for this kind of thing, I review printer/board config mainly :)
> 
> Generally a reviewer does not need to have the actual equipment, but they will need some real world proof that it works. And that can be just from the person submitting a PR. It is not as if we have every printer and every option sent to us for any kind of verification. That would be untenable, as most of us do not do this full time.
> 
> I assume this will be reviewed by Kevin or Arksine, and it just depends when they have time to look at your code and review it.
> 
> Thanks James

@JamesH1978  I am working on getting this setup and working with my setup and I would love to be able to review this and prove that it is working with my board and setup anyway.  Have had mark get a couple things fixed already but i'm still having calibration issues

File: src/endstop.c [hunk 5]
```diff
@@ -17,18 +18,34 @@ struct endstop {
...
+    return gpio_in_read(e->pin);
+#else
+    if(e->type==2)// for Bed Distance sensor
+    {
+        return BD_Data?0:1;
+    }
+    else
+        return gpio_in_read(e->pin);
+#endif
+}
```
Comment (Arksine) [inline src/endstop.c:42]: Much like above, I don't think its valid to put probe specific code here.  I'm not sure what the best software solution for this is, I believe that trigger_sync (trsync.c) can be used?  One short term solution would be to jumper two pins, an input pin the the endstop and an output pin your `bd_sensor.c` sets based on the `BD_Data`.

**Comment by Arksine** [comment]
I added some high level comments on issues that I think need to be addressed before a thorough review can be conducted.   The gist of it is that it will be necessary to move probe/sensor specific code out of the more "generic" modules.  Once that is done I'm willing to look at the module specific code in more depth.

**Comment by markniu** [comment]
> I added some high level comments on issues that I think need to be addressed before a thorough review can be conducted. The gist of it is that it will be necessary to move probe/sensor specific code out of the more "generic" modules. Once that is done I'm willing to look at the module specific code in more depth.

Thank you for those professional comments about the source code, I will try to do that.

File: src/Kconfig [hunk 6]
```diff
@@ -114,6 +114,8 @@ config HAVE_GPIO_SPI
     bool
 config HAVE_GPIO_I2C
     bool
+config HAVE_SOFTWARE_I2C
+    bool
```
[the remark this answers is not in the document]
Comment (markniu) [reply src/Kconfig:118]: Actually, all the MCU support BDsensor with software i2c, but here it cannot pass the git compile check process for some platform that with small rom( compile error: region `rom' overflowed ). that's the reason I add it in each platform.

**Comment by sammcj** [comment]
This would be _amazing_ to have!

**Comment by aivarsV** [comment]
Tried using this branch with bd sensor and had issue with sensor being connected to secondary mcu. https://github.com/aivarsV/klipper/commit/a0fdf9df571273ce34a4f755c53d11f18069630d has changes to make sensor work on secondary mcu. Please consider implementing that.

**Comment by sammcj** [comment]
I have one of these sensors arriving in the next few days - I'm happy to provide feedback / testing if it helps at all.

**Comment by markniu** [comment]
More other features will be added and testing with the BDsensor, so switch to this github: https://github.com/markniu/klipper

**Comment by sammcj** [comment]
@markniu just to confirm - there's no way to use a BD Sensor with klipper unless you run your fork?

**Comment by markniu** [comment]
> @markniu just to confirm - there's no way to use a BD Sensor with klipper unless you run your fork?

There is a solution: run a small code that can add the source code of BDsensor for example BDsensor.py and BD_sensor.c into klipper automatically after git clone from klipper github .
and I will do that recently.

**Comment by Coblenugget1** [comment]
I'm curious if you are planning on adding a second MCU option

**Comment by markniu** [comment]
Now, I think this is the best way to run BD sensor with Klipper by installing patch code into klipper firmware:

```
cd ~
git clone https://github.com/markniu/Bed_Distance_sensor.git
python ~/Bed_Distance_sensor/klipper/install.py 
```

**Comment by sammcj** [comment]
re: 
image


Is the plan to submit a new PR to Klipper to add support @markniu?

I feel like patching Klipper's codebase from the external repo after updates etc... will likely be fraught with potential problems over time (version mismatches, maintenance etc...).

For example I see that the [Python install script](https://github.com/markniu/Bed_Distance_sensor/blob/new/klipper/install_BDsensor_patch.py) does find and replace which means there's not really any error checking or testing taking place as it's not context aware.

**Comment by markniu** [comment]
> re: image
> 
> Is the plan to submit a new PR to Klipper to add support @markniu?
> 
> I feel like patching Klipper's codebase from the external repo after updates etc... will likely be fraught with potential problems over time (version mismatches, maintenance etc...).
> 
> For example I see that the [Python install script](https://github.com/markniu/Bed_Distance_sensor/blob/new/klipper/install_BDsensor_patch.py) does find and replace which means there's not really any error checking or testing taking place as it's not context aware.

yes, you are right, I am trying to make the code more optimization and compatible before PR to Klipper next time.

**Pull request closed without merging**